Wincenttjoi/CS3223-DBMS-Project
# Lab 4

Lab4 Testing:
1) Delete SimpleDBClients/studentdb
2) Run SimpleDBClients/embedded/CreateStudentDB.java (note that I've added creation of index on student.sid)
2) Run SimpleIJ.java from SimpleDBClients.
5) Use queries with join clause. To manually choose which join algorithm is used, modify return result of TablePlanner.makeJoinPlan()

![image](https://user-images.githubusercontent.com/25759159/154648622-49694891-cfde-4e6e-b8de-4aa9d2dcf89c.png)


Note 1: Join on inequality condition not supported, nor is extension 3 (support for (relation, attribute)-pair in queries)
Note 2: If you encounter a BufferUnderflowException after running a lot of tests, repeat step 1-4.

Test cases:

1. select sname, sid, studentid, grade from student, enroll where studentid = sid
2. select sname, sid, grade from student, enroll where 3 = sid
3. select title, deptid, did from course, dept where deptid = did
4. select sname, sid, grade, majorid, did from student, enroll, dept where majorid = did and majorid = did
5. select sname, sid, studentid, grade, majorid, did from student, enroll, dept where studentid = sid and majorid = did
6. select sname, sid, studentid, grade, majorid, did, dname from student, enroll, dept where sid = studentid and grade = 'A' and majorid = did
7. select sname, sid, studentid, grade from enroll, student where sid= studentid and sid = 4
8. select sname, sid, studentid, grade from enroll, student where sid = studentid and studentid = 4
9. select sname, sid, studentid, grade from enroll, student where studentid = sid and studentid = 4
10. select sname, sid, majorid, grade from student, enroll where grade = sname
11. select sname, sid, grade, majorid from student, enroll where majorid = 30

Problems:

- Cases 3, 5, 6, 7 fails with IndexJoin. Case 3 failing is normal because there is no index on the columns selected. Case 5, 6 and 7 gives duplicate results.
- Case 7 should be a result of issue #4
